datafusion/execution/src/object_store.rs

```diff
@@ -89,6 +89,138 @@ pub trait ObjectStoreProvider: Send + Sync + 'static {
     fn get_by_url(&self, url: &Url) -> Result<Arc<dyn ObjectStore>>;
 }
 
+/// Provides a mechanism to get and put object stores.
+pub trait ObjectStoreManager: Send + Sync + std::fmt::Debug + 'static {
+    /// If a store with the same schema and host existed before, it is replaced and returned
+    fn register_store(
+        &self,
+        scheme: &str,
+        host: &str,
+        store: Arc<dyn ObjectStore>,
+    ) -> Option<Arc<dyn ObjectStore>>;
+
+    /// Get a suitable store for the provided URL. For example:
+    ///
+    /// - URL with scheme `file:///` or no schema will return the default LocalFS store
+    /// - URL with scheme `s3://bucket/` will return the S3 store
+    /// - URL with scheme `hdfs://hostname:port/` will return the hdfs store
+    fn get_by_url(&self, url: &Url) -> Result<Arc<dyn ObjectStore>>;
```

Comment:
I am sorry I am missing something -- how is this different than 

https://github.com/apache/arrow-datafusion/blob/860918d17b6bde396b04d718ee1c76d93054bf11/datafusion/execution/src/object_store.rs#L89

?

> While ObjectStoreManager focuses on the polymorphism of how to get an ObjectStore by get_by_url.

It seems like the only function of `ObjectStoreProvider` is a polymorphic `get_by_url` 🤔

Reply:
Thanks @alamb for reviewing. Actually, the main concern is that `ObjectStoreProvider` only covers the `get_by_url`. It does not cover the manual registration. Therefore, in this PR, I introduced `ObjectStoreManager` to cover both. In regard to whether to combine `ObjectStoreManager` and `ObjectStoreProvider`, I have no preference. Here, the main reason of not combining them is for the backward compatibility of existing usage of `ObjectStoreProvider`.

Reply:
`register_store` already exists on `ObjectStoreRegistry` so this would essentially make `ObjectStoreRegistry` and `ObjectStoreProvider` have the same interface which I think is confusing. So as I understand it, the role of `ObjectStoreProvider` is just to allow lazy construction. Manual registration can already be done through the registry. So it just seems to me like we could serve the same use case by holding an `Arc<dyn ObjectStoreRegistry>` in whatever is managing the cache layer.

Reply:
I tried to improve the docs in https://github.com/apache/arrow-datafusion/pull/5577

Reply:
> So it just seems to me like we could serve the same use case by holding an Arc<dyn ObjectStoreRegistry> in whatever is managing the cache layer.

Thanks @thinkharderdev. Since now the `ObjectStoreRegistry` is a struct, I introduced another trait `ObjectStoreManager` for this.